docs/source/notes/numpy_compatibility.rst

```diff
@@ -0,0 +1,388 @@
+   assert(t.is_neg() == False) # Not negative
+
+The :meth:`torch.numpy()` method  and the :doc:`np.asarray()
+<numpy:reference/generated/numpy.asarray>` function returns a **view** of the
+underlying tensor as a ``np.ndarray`` object.
+
+.. doctest::
+
+    >>> b_torch = torch.ones(3)
+    >>> b_torch.numpy()[2] = 32
+    >>> b_torch
+    tensor([ 1.,  1., 32.])
+    >>> a_np = np.array([1, 1, 32], dtype = np.float32)
+    >>> np.array_equal(b_torch.numpy(), a_np) # True
+    True
+    >>> c_tmp = np.asarray(b_torch, dtype = np.float32) # No copy if same dtype
+    >>> c_tmp
+    array([ 1.,  1., 32.], dtype=float32)
+    >>> c_tmp[2] = 1.
+    >>> b_torch
+    tensor([1., 1., 1.])
+
+.. note::
+
+   Since ``np.asarray()`` depends on the implementation of
+   ``torch.Tensor.__array__()`` which calls ``torch.numpy()``, the **leaf node**
+   requirement still needs to be satisfied by the user, i.e., ``requires_grad ==
+   False``
+
+From ``numpy``
+^^^^^^^^^^^^^^
+
+For a NumPy ``ndarray`` to be convertible to a PyTorch tensor:
+
+- It must have only native byte order
+- Must not be read only
+- Array strides must be multiples of the Torch element byte size
+- Must have a ``dtype`` which is one of ``float64 float32 float16 complex64
+  complex128 int64 int32 int16 int8 uint8 and bool``
+
+Concretely, these may be expressed as:
+
+.. doctest::
+
+   >>> a_np = np.ones(4).reshape(2, 2)
+   >>> b_torch = torch.tensor(a_np)
+   >>> assert(a_np.dtype.byteorder == '=') # Native byte order
+   True
+   >>> assert(a_np.flags.writeable == True) # Not read only
+   True
+   >>> np.equal([stride % b_torch.element_size() for stride in a_np.strides], np.zeros(len(a_np.strides))) # Multiples of torch element byte size
+   array([ True,  True])
+   >>> a_np.dtype in ["float64", "float32", "float16", "complex64", "complex128", "int64", "int32", "int16", "int8", "uint8", "bool"] # Supported dtype
+   True
+
+To obtain a **view** of the data, :meth:`torch.from_numpy()` can be used.
+
+.. doctest::
+
+   >>> a_np = np.array([1, 2, 3], dtype = np.float64)
+   >>> b_torch = torch.from_numpy(a_np)
+   >>> # b_torch = torch.as_tensor(a_np) # see note
+   >>> b_torch[2] = 23
+   >>> b_torch
+   tensor([ 1.,  2., 23.], dtype=torch.float64)
+   >>> a_np[0] = 22
+   >>> b_torch # view, changes with a_np
+   tensor([ 22.,  2., 23.], dtype=torch.float64)
+   >>> np.array_equal(b_torch.numpy(), a_np)
+   True
+
+- :meth:`torch.from_numpy()` is guaranteed to share memory with NumPy.
+- :meth:`torch.as_tensor()` will try to stay away from copy operations, it
+  also has the effect of sharing memory. However, ``torch.as_tensor()`` has
+  slightly higher overhead as it checks and accepts other iteratable objects as
+  well, e.g. ``list`` objects.
+- :meth:`torch.from_dlpack()` called with a NumPy array (``np.version.version >=
+  1.20``) as its argument will also generate a ``torch.Tensor`` view.
+
+To obtain a **copy** of the ``ndarray`` object, and not share memory, the
```

Comment:
Alternatively the user can explicitly make a copy after the conversion to a tensor